dotnet-test-samples/async-architectures/async-lambda-dynamodb/src/AsyncTesting.S3EventHandler/Function.cs

```diff
@@ -0,0 +1,111 @@
+{
+    private IAmazonS3 _s3Client;
+
+    /// <summary>
+    /// Default constructor. This constructor is used by Lambda to construct the instance. When invoked in a Lambda environment
+    /// the AWS credentials will come from the IAM role associated with the function and the AWS region will be set to the
+    /// region the Lambda function is executed in.
+    /// </summary>
+    public Function() : this(null)
+    {
+        AWSSDKHandler.RegisterXRayForAllServices();
+    }
+
+    /// <summary>
+    /// Constructs an instance with a preconfigured S3 client. This can be used for testing outside of the Lambda environment.
+    /// </summary>
+    /// <param name="s3Client"></param>
+    public Function(IAmazonS3? s3Client)
+    {
+        this._s3Client = s3Client ?? new AmazonS3Client();
+    }
+
+    /// <summary>
+    /// This method is called for every Lambda invocation. This method takes in an S3 event object and can be used 
+    /// to respond to S3 notifications.
+    /// </summary>
+    /// <param name="evnt"></param>
+    /// <param name="context"></param>
+    /// <returns></returns>
+    [Logging(LogEvent = true)]
+    [Tracing]
+    public async Task FunctionHandler(S3Event evt)
+    {
+        var eventRecords = evt.Records ?? new List<S3Event.S3EventNotificationRecord>();
+        
+        Logger.LogInformation($"Received {eventRecords.Count} record(s) from S3");
+
+        foreach (var record in eventRecords)
+        {
+            await this.ProcessS3Record(record);
+        }
+    }
+
+    [Tracing]
+    private async Task ProcessS3Record(S3Event.S3EventNotificationRecord record)
+    {
+        var s3Event = record.S3;
+
+        if (s3Event == null)
+        {
+            Logger.LogWarning("S3 Event object is null, skipping");
+            return;
+        }
+
+        try
+        {
+            Logger.LogInformation($"Processing object {record.S3.Object.Key} from bucket {record.S3.Bucket.Name}");
+
+            var originalObject = await this._s3Client.GetObjectAsync(
+                record.S3.Bucket.Name,
+                record.S3.Object.Key);
+
+            // convert stream to string
+            var reader = new StreamReader(originalObject.ResponseStream);
+
+            var lowerCaseMessage = await reader.ReadToEndAsync();
+
+            Logger.LogInformation($"Parsed event body is {lowerCaseMessage}");
+
+            var upperCaseMessage = lowerCaseMessage.ToUpper();
+
+            var inputBytes = Encoding.ASCII.GetBytes(upperCaseMessage);
+
+            Logger.LogInformation("Storing transformed data in S3");
+
+            await this._s3Client.PutObjectAsync(
+                new PutObjectRequest()
+                {
+                    BucketName = Environment.GetEnvironmentVariable("DESTINATION_BUCKET"),
+                    InputStream = new MemoryStream(inputBytes),
+                    Key = record.S3.Object.Key
+                });
+        }
+        catch (Exception e)
+        {
```

Comment:
Is there a flow that would cause an exception? is it tested? or do we prefer to "crash" the lambda and get the log that way (not sure myself)

Reply:
It's a good question. I think if I was doing this in the real world I'd actually introduce some kind of DLQ. I'll add that to the template to demonstrate.